server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.AbstractDiffable;
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+public final class DataStream extends AbstractDiffable<DataStream> implements ToXContentObject {
```

Comment:
Rename to `DataStreamMetaData` (like `IndexMetaData`)?

Reply:
Data stream definitions will be stored in a custom metadata (like `ComponentTemplateMetadata`). These classes tend to have the `Metadata` suffix and therfore I think that adding `MetaData` suffix is confusing here. 

Maybe we can rename this to `DataStreamSource`? Like how stored scripts are stored? (which also is stored inside a custom metadata).

Reply:
OK. I think I imagined a more native integration into the `MetaData` (outside `Custom`). I think the primary purpose of Custom is for extensibility (x-pack/plugins/modules)? In particular, I think `MetaData.getAliasAndIndexLookup` will need some support of data-streams? It could still look for the `Custom` metadata pieces, but feels second-class.

Let us leave this as is for now and tackle above when we get to add the metadata and do the lookups.

Reply:
> I think MetaData.getAliasAndIndexLookup will need some support of data-streams?

We can add a getter on metadata that reads data stream from a custom and then it is as if it is stored as a primary field. The reason why I think we should do this is that, we get serialization, diffability and versioning for free without changing the Metadata class. 

> It could still look for the Custom metadata pieces, but feels second-class.

Perhaps we should rename this in the future. Scripts and pipelines and other primary concepts are stored also as custom metadata and these concepts aren't second class either.